Enable perfect forwarding with irept/exprt/typet constructors

tautschnig opened the pull request:
This should enable use (with performance benefit) of rvalue references in
higher-level APIs.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

tautschnig committed:
Pass irept by value in modifying irept::add/set operations

This should enable use (with performance benefit) of rvalue references in
higher-level APIs.

peterschrammel reviewed (commented):
> n/a My commit message includes data points confirming performance improvements (if claimed)

Isn't your commit message claiming exactly that? ;)

tautschnig commented:
> Isn't your commit message claiming exactly that? ;)

No, I'm not yet claiming that: I'm working on a PR that will actually start making serious use of these functions, at which point I should likely do a proper evaluation.

tautschnig commented:
Although I have a fix for the test failures I'm marking this do-not-merge as I'll add the corresponding changes to *typet/\*exprt/code\*t in a single branch to do performance benchmarking. (Also, the failing tests did confirm that the functions _are_ used.)

tautschnig committed:
Add pass-by-value constructor to exprt

Uses the recently added irept constructor to move all arguments in place.

tautschnig committed:
Pass-by-value constructors for typet and type_with_subtypet

Also don't use subtype() in the constructor as that would unnecessarily
construct a blank irept first.

kroening commented on src/util/irep.cpp:
```diff
@@ -346,6 +346,32 @@ irept &irept::add(const irep_namet &name, const irept &irep)
+
+  if(!entry.second)
+    entry.first->second = std::move(irep);
+
```
Does the above work? I'd be worried that the first std::move gets evaluated, and then second call doesn't have anything to move anymore.

tautschnig replied:
No, it did not. I've got an implementation that does, will push soon - just working on PRs that set up proper use of all this.

tautschnig commented:
Rvalue constructors added. I will now start testing and benchmarking. Various preparatory changes have been factored out into other PRs, and there is a lot that could and should be done on top of this one.

smowton commented on src/util/irep.cpp:
```diff
@@ -346,6 +346,32 @@ irept &irept::add(const irep_namet &name, const irept &irep)
+
+  return it->second;
+#else
+  named_subt::iterator entry = s.find(name);
```
How about just `s[name] = std::move(irep);`?

tautschnig replied:
We also want to return a reference to the newly added `irept`.

smowton commented on src/util/std_code.h:
```diff
@@ -479,6 +494,11 @@ class code_assumet:public codet
     add_to_operands(expr);
   }
 
+  explicit code_assumet(exprt &&expr):codet(ID_assume)
```
Style: spacing of `:` (here and below)

smowton commented:
Paging @NathanJPhillips for C++ wizardry: If you have a function with overloads `f(const xyz &)` and `f(xyz &&)`, and a function that takes a few xyzs and applies `f` to each one, is there a good way to get optimal use of `f`? For example, in this PR we have something like

```
g(const xyz &a, const xyz &b, const xyz &c)
{
  f(a); f(b); f(c);
}

g(xyz &&a, xyz &&b, xyz &&c)
{
  f(std::move(a)); f(std::move(b)); f(std::move(c));
}
```

But, (a) it's annoying to duplicate every method like `g`, which here means all the constructors on `exprt` children, and (b) this doesn't catch the case where a, b and c are mixed lvalue and rvalue refs.

I'd like a case like
```
template<class A, class B, class C>
g(A &&a, B &&b, C &&c)
{
  f(std::forward(a)); f(std::forward(b)); f(std::forward(c));
}
```

But it seems odd to have A, B and C be that general when actually I only want to allow `const xyz &` or `xyz &&`. Is there a nice way to do this?

martin-cs commented on src/util/irep.h:
```diff
@@ -21,6 +21,7 @@ Author: Daniel Kroening, kroening@kroening.com
 // #define HASH_CODE
 #define USE_MOVE
 // #define NAMED_SUB_IS_FORWARD_LIST
+#define IS_CANONICAL_BIT
```
I have feelings on this subject and I feel compelled to share them with you.

1. Before this can be merged we *need* to add the assertion to the simplifier that checks idempotency.  By design it should hold but ... I've been known to bet people up to 50p that it doesn't in one or more obscure cases.

2. We should also think about the cache in the simplifier.

3. Crazy idea : rather than just a bit, we could make simplification just a bit more union-find and have a pointer to the canonical version.  This would allow fast simplification and deduplification / re-introduction of sharing.

4. This road starts to feel like it leads to canonical irept's.  I remember discussing this ages ago and coming to the conclusion that this was common in SMT solvers because their nodes were simple and unchanging while CPROVER really needed them to be able to gain annotations / fields.  If so this could be ... a thing.

tautschnig committed:
Pass-by-value in *exprt constructors

Profiling does not show any advantage of using rvalue constructors over passing
by value (and then moving in place), and this requires at most as many copies as
the previous const-reference case. Also make use of the new exprt constructors
to set up operands without an additional add_to_operands call.

tautschnig committed:
Pass-by-value in *typet constructors

Profiling does not show any advantage of using rvalue constructors over passing
by value (and then moving in place), and this requires at most as many copies as
the previous const-reference case.

tautschnig committed:
Pass-by-value in code*t constructors

Profiling does not show any advantage of using rvalue constructors over passing
by value (and then moving in place), and this requires at most as many copies as
the previous const-reference case. Also make use of the new exprt constructors
to set up operands without an additional add_to_operands call.

tautschnig commented:
After a lot of back&forth in this branch to play with several optimisations this PR now does what the $SUBJECT says. Profiling is still running to make sure I haven't messed  up, but reviews would be appreciated.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f9bfdb0cb645d945e6358595a23c2ff9be7ccc2b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100651205
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e224f00a5c1b0853a7dbbe7f018ea9c25225891f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100654772

tautschnig commented:
No obvious regressions in my profiling results thus far, but now it's also possible to identify some offenders (first candidate: `object_descriptor_exprt()`). I'll keep reviewing the profile and will add small changes to this PR as needed.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d272452bf5c4b49becbd350aaa6ce12d630f1829).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100672863

kroening commented:
> Paging @NathanJPhillips for C++ wizardry: If you have a function with overloads `f(const xyz &)` and `f(xyz &&)`, and a function that takes a few xyzs and applies `f` to each one, is there a good way to get optimal use of `f`? For example, in this PR we have something like

For data structures that have an efficient move constructor it's hardly ever useful to have both the `const &` and the `&&` variant -- passing by value is the most efficient answer. If the thing being passed is stored, then use pass by value plus `std::move`, if it's just "inspected", then use "const &". In particular, this avoids the need to contemplate the cross-product of `const &` and `&&` when multiple arguments are given.

kroening commented:
As example, the program
```
class data
{
public:
  data() { }
  data(const data &) { }
  data(data &&) { }
};

void func(data d1)
{
  data d2 = std::move(d1);
}

int main()
{
  func(data());
}
```
never actually calls that copy constructor.

tautschnig commented:
Just for the record: @NathanJPhillips had shared his knowledge with me offline back in December, which is completely in line with what @kroening said above. (And profiling confirms that the compiler is doing what is expected of it.)

tautschnig commented:
I'll keep benchmarking and adding minor nudges to this as I find them in the profiling data, but otherwise this is looking ok from a performance point of view. We should certainly start thinking about factories that construct expressions like `true_exprt` just once (I'll know in a few hours whether this is now a bit better, but the data from this morning suggested that we spend 200 seconds on 5B `true_exprt` constructions (in my current SV-COMP/ReachSafety-ECA set)).

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d871239a5aae1703cb60a467b4fca8b0cf2da4da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100949504
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c58107c794bcf028538ea1a86f22b8fa906d4b9b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100955074

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 77d73c1ece8d8831f9dcea68eb5bd253f638ff8d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101066127

kroening commented on src/util/expr.h:
```diff
@@ -59,9 +59,18 @@ class exprt:public irept
+  exprt(irep_idt _id, typet _type, operandst &&_operands)
+    : irept(
+        std::move(_id),
+        {{ID_type, std::move(_type)}},
```
moving a `dstring` may well be slower than copying it.

tautschnig replied:
Will remove.

tautschnig replied:
Or actually: shouldn't we just define a move constructor for `dstringt` that actually is cheap? For the case that `irep_idt == std::string` we'd still want the `std::move`.

tautschnig replied:
I have added a commit that adds efficient move construction/assignment to `dstringt`.

kroening replied:
I am still not entirely convinced that we need those `std::move(id)` -- we wouldn't do it for an integer, say.

tautschnig replied:
But we don't know whether `irep_idt` is a `std::string` (where it would seem worthwhile) or a `dstringt` (where it's certainly unnecessary). But also that isn't something I want this PR to get blocked on, I'm happy to follow whatever I'm advised to do.

kroening commented on src/util/irep.h:
```diff
@@ -440,9 +441,15 @@ class irept
+  {
+    add(name, std::move(irep));
+  }
   void set(const irep_namet &name, const long long value);
```
Since this will copy eventually, simply have one variant, with just `irept`, and then move that?

tautschnig replied:
Done.

kroening commented on src/util/irep.h:
```diff
@@ -424,6 +424,7 @@ class irept
   irept &add(const irep_namet &name);
   irept &add(const irep_namet &name, const irept &irep);
+  irept &add(const irep_namet &name, irept &&irep);
 
```
One variant with just `irept` suffices.

tautschnig replied:
Done.

kroening commented:
You sometimes use the `.swap(other)` and sometimes the `= std::move(other)` pattern -- this will ultimately produce the same code, but it may nevertheless be nice to stick to one pattern. I suppose the `std::move` variant is the one that's more modern.

tautschnig commented:
> You sometimes use the .swap(other) and sometimes the = std::move(other) pattern -- this will ultimately produce the same code, but it may nevertheless be nice to stick to one pattern. I suppose the std::move variant is the one that's more modern.

I used to have a commit that would clean this up across the codebase, but eventually got tired of maintaining it. I'll at least clean up those bits that affect this PR.

tautschnig committed:
Move constructor/assignment for dstringt

These do not actually destroy the source object, but ensure that using dstringt
with std::move is as efficient as a regular copy/assignment.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 8cc4c5eabcc85eaec99065bf5972b39ce4292d11).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103558660
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

kroening commented on src/util/ssa_expr.h:
```diff
@@ -18,7 +18,7 @@ class ssa_exprt:public symbol_exprt
   /// \param expr: Expression to be converted to SSA symbol
-  explicit ssa_exprt(const exprt &expr) : symbol_exprt(irep_idt(), expr.type())
+  explicit ssa_exprt(const exprt &expr) : symbol_exprt(expr.type())
   {
```
I kind of liked that `symbol_exprt` now enforces that you set the identifier.

tautschnig replied:
I don't think `symbol_exprt` actually does? This change here just removes an unnecessary call to `set_identifier`.

kroening replied:
No, it doesn't, but I was hoping it would eventually.

tautschnig replied:
We can certainly put the `irep_idt()` back if and when we do.

kroening commented on src/util/std_code.h:
```diff
@@ -2104,17 +2077,17 @@ class side_effect_expr_function_callt:public side_effect_exprt
   {
-    arguments() = _arguments;
+    arguments() = std::move(_arguments);
   }
```
You could make the above `multi_ary_exprt(ID_arguments, std::move(_arguments))`

tautschnig replied:
Done.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 4b20ea0a1e72840e180180d1191460fe9bc1f1c3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104425235
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

kroening commented on src/util/std_expr.h:
```diff
@@ -3194,10 +3165,9 @@ inline address_of_exprt &to_address_of_expr(exprt &expr)
   {
-    PRECONDITION(op.type().id()==ID_bool);
+    PRECONDITION(op().type().id() == ID_bool);
   }
```
Good catch

tautschnig replied:
~Hmm, but then I should really add `as_const()` in there ;-)~ Edit: Done.

kroening replied:
Unfortunately, this doesn't do the job -- the mere access to `operands()` is enough to wreck the sharing.
This needs to be `as_const(*this).type().id()`.

tautschnig replied:
My apologies for failing to take into account this remark right here, the fix is now in #4391.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0520791d000f596d875001a97607f169441cccd1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104426210
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 83ed9a5acb93e1546166e0e0b754ea7dd572abfa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104448810
